core/src/main/java/io/parsingdata/metal/expression/value/reference/Ref.java

```diff
@@ -48,24 +52,36 @@
  */
 public class Ref<T> implements ValueExpression {
 
-    public final T reference;
-    public final Predicate<ParseValue> predicate;
+    public final ImmutableList<T> references;
+    public final BiPredicate<ParseValue, T> predicate;
     public final SingleValueExpression limit;
 
-    private Ref(final T reference, final Predicate<ParseValue> predicate, final SingleValueExpression limit) {
-        this.reference = checkNotNull(reference, "reference");
+    @SafeVarargs
+    private Ref(final BiPredicate<ParseValue, T> predicate, final SingleValueExpression limit, final T reference,  final T... references) {
         this.predicate = checkNotNull(predicate, "predicate");
         this.limit = limit;
+        this.references = create(checkContainsNoNulls(references, "references"))
+            .add(checkNotNull(reference, "reference"));
+    }
+
+    private Ref(final BiPredicate<ParseValue, T> predicate, final SingleValueExpression limit, final ImmutableList<T> references) {
```

Comment:
This constructor is meant as a copy constructor to avoid converting an immutable list to an array. In combination with the `withLimit` method below, this constructor can be kept private. The `withLimit` is a copy method used in the Shorthand when ref is used in combination with last.